### ISSUE TITLE
[feature] How to define dependency on an individual package component(s) in a conanfile.py to avoid leaking transient dependencies?

### ISSUE THREAD

**Issue by 0x4d0x4b** [conan-io/conan#7475]
I have got a package containing pre-built libraries provided by a 3rd party vendor. There are two libraries inside of a package that are mutually exclusive. Linking to both of them at the same time is undefined behavior.
For the sake of an example:
```
├── vendor_protocol
    ├── all
        ├── conanfile.py
        ├── include
        │   └── protocol_api.h
        ├── lib
            ├── libfullapi.so
            └── libliteapi.so
```
 
The `conanfile.py` content for `vendor_protocol` package is
```
from conans import ConanFile, tools

class InvisionConan(ConanFile):
    name = "vendor_protocol"
    version = "1.2.3"
    # only os since we have got prebuilt libraries
    settings = "os"
    exports_sources = "include/*", "lib/*"

    def configure(self):
        if self.settings.os != "Linux":
            raise Exception("Only linux compatible")

    def package(self):
        self.copy("include/*")
        self.copy("lib/*")

    def package_info(self):
        self.cpp_info.components["lite"].libs = ["liteapi"]
        self.cpp_info.components["full"].libs = ["fullapi"]
```	
On top of `vendor_protocol` package lies our wrapper layer - in this example `mylib`
In that wrapper package `cmake_find_package` generator is used (plain `cmake` generator does not seem to provide necessary targets referring to conan components)
`CMakeLists.txt` is very simple and allows use of `vendor_protocol::full` target without any issue.
So far so good.
```
cmake_minimum_required(VERSION 3.14)

project (mylib CXX)

set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/lib)
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/lib)

find_package(vendor_protocol)

add_library(${PROJECT_NAME} ${CMAKE_CURRENT_SOURCE_DIR}/src/wrapper.cpp)
target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)

if(TARGET vendor_protocol::full)
    target_link_libraries(${PROJECT_NAME} PUBLIC vendor_protocol::full)
else()
    message(FATAL_ERROR "vendor_protocol::full not found")
endif()

install(TARGETS ${PROJECT_NAME})
```

The problems start when we look at `conanfile.py`

```
from conans import ConanFile, CMake, tools


class MylibConan(ConanFile):
    name = "mylib"
    version = "0.1"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False]}
    default_options = {"shared": True}
    generators = "cmake_find_package"
    exports_sources = "CMakeLists.txt", "src/*", "include/*"

    requires = ["vendor_protocol/1.2.3"]

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def package(self):
        self.copy("lib/*")
        self.copy("include/*")

    def package_info(self):
        self.cpp_info.libs = ["mylib"]
```
		
If I left `package_info(self)` in `mylib` like presented above and defined a dependency on the library in a consumer application by linking to `mylib::mylib` like in the `CMakeLists.txt` below, unnecessary transient dependencies would be leaked into a project i.e. the app would be linking both `libfullapi.so` and `libliteapi.so` from `vendor_protocol` which must not happen.
```
cmake_minimum_required(VERSION 3.14)

project (myapp CXX)

set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin)

find_package(mylib)

add_executable(${PROJECT_NAME} ${CMAKE_CURRENT_SOURCE_DIR}/src/main.cpp)

if(TARGET mylib::mylib)
    target_link_libraries(${PROJECT_NAME} PRIVATE mylib::mylib)
else()
    message(FATAL_ERROR "mylib::mylib not found")
endif()

install(TARGETS ${PROJECT_NAME})
```

To get expected behavior I had to modify `package_info(self)` in `mylib`'s `conanfile.py` in a following way
```
    def package_info(self):
        # no way for the entire package to depend on individual component(s)
        # it would be welcomed if one could simply define
        # self.cpp_info.libs.requires = ["vendor_protocol::full"]
        #
        # instead lite component has to be filtered out
        libs_to_discard = self.deps_cpp_info["vendor_protocol"].components["lite"].libs
        for lib in libs_to_discard:
            self.deps_cpp_info["vendor_protocol"].libs.remove(lib)

        del self.deps_cpp_info["vendor_protocol"].components["lite"]

        # another way would be to define component in this library and then set
        # requires for it like
        # self.cpp_info.components["dummy"].requires = ["vendor_protocol::full"]
		# and link to target mylib::dummy in a cmake list
        # but why defining a component if it is not needed really
        #
        # it is not possible to define component with the same name as package either
		# as it is restricted by conan itself
        # so it is not allowed to define self.cpp_info.components["mylib"].requires to
        # take advantage of mylib::mylib target later
        self.cpp_info.libs = ["mylib"]
```
		
What is correct way to use conan components in that context? Is my workaround the only way at the moment?
I do not want to split `vendor_protocol` into two single-library packages and I do not want to introduce `dummy` component in the wrapper library `mylib`.

- [x ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
From https://github.com/conan-io/conan/issues/7496 too, it seems this would be a valid feature request, to have something like:

```python
self.cpp_info.requires = ["liba::requester"]
```

**Comment by fulara** [comment]
@memsharded  Just a note here, this is breaking change for components.
I think the ideal behavior would be that the packages that provide components do not propagate automaticlaly, but have to be put into a `cpp_info.requires[...]`
With an assert preferably when someone does not do that - as in this case it would not addanything - rather than whats the current behavior seemingly that it propagates everything.

**Comment by jgsogo** [comment]
IMHO, a package should ALWAYS be usable without components (some generators might not provide something like a _components_ feature). If two components are mutually exclusive then the way to go is to add an option to the recipe and generate the `lite` or `full` flavor (and create only the matching component if desired)

Said that, I agree that this feature could be useful and I agree with @fulara about making it explicit (but not only for packages that provide components, but for all of them). This would be a breaking change, it would require rewriting all the recipes... I think it is important that if requires need to be declared explicitly when there are components, they should be declared too when no components are used. This would be a big change in the way Conan works, and I'd need to play with it to know if I'd be comfortable with it as a developer.

**Comment by fulara** [comment]
Just a comment here, so we dont get off tracked here. 
This request should no longer focus about handling the mutually exlusive packages, 
its only so that component in my mind fulfil the promise they seem to give.

Example:
we have boost, which has tons of libraries, currently they are all linked.
I would like to see componentized boost, like boost::system.

And I would like all consumers to explicitly specify on which particular component they depend, when they boost::system on their list, this is the only library that should be carried towards that particular dependency tree.

Currently every dependency will be carried if the consumers dont use components - which seems counter intuitive for me. 

Note: This should only affect consumers of componentized packages.

**Comment by 0x4d0x4b** [comment]
Yes, the bottom line is that a user should be allowed to define dependency on a subset of components (link to a subset of libraries) from a package and only that subset should be carried down the dependency graph.
I used an example with mutually exclusive libraries to highlight that linking to everything might be an issue and it is my real life example (simplified a bit, but real). 
Indeed, in case of 2 mutually exclusive libraries the issue can be resolved using an option, however that would not be the case for bigger number of libraries as the number of available combinations would explode.
I agree with @fulara , let's focus on a generic case of multi-library package being consumed and only subset of libraries/components used.

**Comment by jgsogo** [comment]
Current situation is:
 * using components: **nothing** is required unless explicit: `self.components["xxx"].requires["req::libA"]`
 * without components: everything is linked

Possible feature for Conan v1.x:
 * using components: idem
 * without components: **everything** is linked unless explicit: `self.cpp_info.requires = ["req::libA"]`

I don't like the different behavior with/without components (harder to reason about, harder to document), but I understand that using components it is nice to be explicit.... and if not using components, I don't like forcing the user to declare the requires once again (they are already listed as `requires = ....`) (maybe it is just because I'm not used to it).

We need to take into account that recipes from Conan Center will be used from Conan v1.x and Conan v2.0 for a long time and they need to work using both versions.

**Comment by fulara** [comment]
@jgsogo  please see: https://github.com/conan-io/conan/issues/7496

components only fulfil their promise if everyone up in the hierarchy has used this  line `syntax:self.components["xxx"].requires["req::libA"]` 
**Otherwise everything is linked.** 

>using components: nothing is required unless explicit: self.components["xxx"].requires["req::libA"] 
> without components: everything is linked

 that is not true. ( as shown in my example code. )

**Comment by jgsogo** [comment]
Let me elaborate your example here, I need to see it with an example:

your `libA`:

```python
class Recipe(ConanFile):
    name = "liba"

    def package_info(self):
        self.cpp_info.components["publisher"].libs = ["publisher"]
        self.cpp_info.components["requester"].libs = ["requester"]
```

Then you are consuming `libA`:

```python
class Recipe(ConanFile):
    name = "consumer"
    requires = "liba/..."

    def package_info(self):
        self.cpp_info.libs = ["libb"] # because i am using cpp_info.libs conan is causing dependence on whole liba (all libs from it. )
        #self.cpp_info.requires = ["liba::requester"] #i wish this line was working as expected.
```

but your CMake scripts would use something like this:

```CMake
find_package(liba)

add_library(libb ...)
target_link_libraries(libb PUBLIC liba::requester)
```

So, up to this first-level consumer, you are using only the component you want to use.


**The problem starts with transitive consumers:**

```python
class Recipe(ConanFile):
    name = "transitive_consumer"
    requires = "libb/..."
```

and here you can only link with target `libb::libb`, which in turn links with `liba::liba` (all the components)

**Comment by fulara** [comment]
@jgsogo  the problem starts where i am consuming 'liba' - that package_info is wrongly representing the dependencies. 
It shouldnt require on all the packages but for the case of component it should only depend on the ones i .require.

This doesnt work like that, hence the proper usage of components *forces* usage of compnents up the stack.
And to add to that what @kajdam  mentioned above, is that you have to specify some dummy name because there is an assert. 

If I am correct about my understanding if I had a flow:
liba <- libb <- libc <- bin
Then in order to achieve the promise of component (where from liba I would want to use ::requester) just because of that I am forced to use components in libb libc! and these packages dont want to specify components - they want to expose a single library. 

Your documentation states that components are oriented towards things like boost, and this is exactly what people want to have when using boost - they only want to depend onboost::system and only that would be passed to linker, up the whole stack.

**Comment by jgsogo** [comment]
Yes, I totally understand your point and I really think this is something we should improve. Your proposal with:

```python
class Recipe(ConanFile):
    name = "consumer"
    requires = "liba/..."

    def package_info(self):
        self.cpp_info.libs = ["libb"] # because i am using cpp_info.libs conan is causing dependence on whole liba (all libs from it. )
        self.cpp_info.requires = ["liba::requester"] #i wish this line was working as expected.
```

it looks like the way to go, the recipe declares that `libb::libb` should link with `liba::requester`. Then, transitive consumers using `libb::libb` will use nothing more than the `requester` from `liba`.

**Comment by fulara** [comment]
@jgsogo  great :)!
Just last points from my pov.

If one introduces a component on his package the default behavior conan has taken so far was to only behave in 'component' like fashion if consumers of that package has used component as well, its not something I would go with personally but I do understand why this was made.  
As we are now used to produce non componenti packages, usually only packages that we would like to support component are big third party libraries, such as boost. 

Now, even if we introduce component to boost, there will be no gain from it, unless we change all the recipes everywhere to make use of it.
And in my simple minded opinion we should only *depend on what we need* - so recipes ought to be analyzed and they should depend only on libraries that they require, for boost even probably a lot of cases out there wont require any dependencies other than header files.

So in the end I see two solutions to this, both of them assume introduction of the `requires` or `requires_component` ( which is a mouthful )
1) maintain current behavior however make it so that if one uses `requires` - it changes the way dependencies works ( in a way that it clears other libs from that come from that dependency. ) - this would not break any existing recipes, and at first wont provide much benefits from components, recipes will over time align to those. 
2) What I wrote above, which is to completely break current version of the component, and make it so that they only affect your package if you have put them explicitly in the package_info. 

In perfect world I would personally go for 2) I would possibly introduce a keyword called 'full' in cases where one does not care. such as .requires['boost::full'] or .requires["full']

If the 2nd solution would be chosen, I would love an assert ( there alread exists such asserts if one enables use of components and depends on component package but forgot to use any .requires ).

in 1) option I would possibly consider an option to warn on using component like this, but probably not by default.


Thanks for this feature anyway, I learned about this feature by surprise and it actually is something I've given up on a long ago. ( I 've learned to live with conan packages be 'all or nothing'. )

**Comment by jgsogo** [comment]
There is one thing to consider while implementing this feature: at the moment of creating the consumer recipe, the requirement provides components and that's the reason why we could use something like `self.cpp_info.requires = ['req::req_component']`. Nevertheless, give then different mechanisms that Conan provides to override recipes and versions, in an actual graph the requirement could use an older version without components and we cannot break here (do we agree?).

It means that the consumer need to check if the targets exist (`req::req_component`) before using it with `target_link_components` as a transitive dependency, and, if it doesn't, it should fallback to link with the full component: `req::req`.

Much better: it should check if the requirement provides any component at all to decide what to link, otherwise if the recipe author writes a typo they wouldn't notice about the error. This variable could be really useful: https://github.com/conan-io/conan/blob/362e81ca1e4ef4d7b3aaca225c653a70e7c641ce/conans/client/generators/cmake_find_package.py#L71

**Comment by memsharded** [comment]
This has been added, using the ``self.cpp_info.requires`` syntax, it will be possible to declare dependencies on components, even if the current package is not defining components. Will be released in Conan 1.30.
